torch_xla/experimental/xla_sharding.py

```diff
@@ -1,14 +1,74 @@
+from collections import OrderedDict
 import torch
 import torch_xla
 import torch_xla.core.xla_model as xm
 from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
+from torch_xla.experimental.pjrt import requires_pjrt
 
 import numpy as np
-from typing import Tuple, Union
+from typing import Tuple, Union, List
 
 
-def mark_sharding(t: Union[torch.Tensor,
-                           XLAShardedTensor], mesh_shape: Tuple[int],
+class Mesh:
+  """Describe the logical XLA device topology mesh and the underlying resources.
+
+  Args:
+    device_ids (Union[np.ndarray, List]): A raveled list of devices (IDs) in a custom order. The list is reshaped
+        to an `mesh_shape` array, filling the elements using C-like index order.
+
+    mesh_shape (Tuple[int, ...]): A int tuple describing the logical topology shape
+        of the device mesh, and each element describes the number of devices in
+        the corresponding axis.
+
+    axis_names (Tuple[str, ...]): A sequence of resource axis names to be assigned to the dimensions
+        of the `devices` argument. Its length should match the rank of `devices`.
+
+  Example:
+  —------------------------------
+  mesh_shape = (4, 2)
+  num_devices = len(xm.get_xla_supported_devices())
+  device_ids = np.array(range(num_devices))
+  mesh = Mesh(device_ids, mesh_shape, ('x', 'y'))
+  mesh.get_logical_mesh()
+  >> array([[0, 1],
+            [2, 3],
+            [4, 5],
+            [6, 7]])
+  mesh.shape()
+  >> OrderedDict([('x', 4), ('y', 2)])
+  """
+
+  device_ids: np.ndarray
+  mesh_shape: Tuple[int, ...]
+  axis_names: Tuple[str, ...]
+
+  def __init__(self,
+               device_ids: Union[np.ndarray, List],
+               mesh_shape: Tuple[int, ...],
+               axis_names: Tuple[str, ...] = None):
```

Comment:
Just curious - how will `axis_names` be used long-term? Is it just for annotating the mesh?

Reply:
Good question, mesh axis annotation is useful since it makes the annotation logic more readable. We can also build a partitioning rule based on the axis name, instead of int indices.